triopg/_tests/test_triopg.py

```diff
@@ -174,9 +174,90 @@ def _listener(connection, pid, channel, payload):
     await triopg_conn.remove_listener("foo", _listener)
     assert await asyncpg_execute("NOTIFY foo, '3'")  # Should be ignored
 
-    with trio.fail_after(1):
+    with trio.fail_after(5):
         channel, payload = await listener_receiver.receive()
     assert channel == "foo"
     assert payload == "2"
     with pytest.raises(trio.WouldBlock):
         await listener_receiver.receive_nowait()
+
+
+@pytest.mark.trio
+async def test_listener_cancel(triopg_conn, asyncpg_execute):
```

Comment:
It might be good to explain a bit further what this test aims at.

From what I understand, the point is `cancel_scope.cancel` is a synchronous function that doesn't raise any `trio.Cancelled` exception. So once called, the next statements are executed until an await occurs which should be where the `trio.Cancelled` is raised (considering the await actually reaches the trio loop of course)
So the given `remove_listener` is an async function, we need to make sure it doesn't start by doing some synchronous stuff, before being cancelled when doing an asynchronous operation, hence ending up in an inconsistent state.

Reply:
Thank you for explanation! Cancellation is weird :) I've added explanation and `assert_listeners`, that also checks Postgres/asyncpg state consistency.

Reply:
Good idea to isolate the use of `conn._asyncpg_conn._listeners` given it is not public ;-)